File: app/src/main/java/it/chalmers/gamma/adapter/primary/web/SuperGroupController.java
package it.chalmers.gamma.adapter.primary.web;

import it.chalmers.gamma.app.supergroup.SuperGroupFacade;
import it.chalmers.gamma.app.supergroup.domain.SuperGroupRepository;
import org.springframework.stereotype.Controller;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.servlet.ModelAndView;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Controller
public class SuperGroupController {

    private final SuperGroupFacade superGroupFacade;

    public SuperGroupController(SuperGroupFacade superGroupFacade) {
        this.superGroupFacade = superGroupFacade;
    }

    @GetMapping("/super-groups")
    public ModelAndView getSuperGroups(@RequestHeader(value = "HX-Request", required = false) boolean htmxRequest) {
        List<SuperGroupFacade.SuperGroupDTO> superGroups = this.superGroupFacade.getAll();

        ModelAndView mv = new ModelAndView();
        if(htmxRequest) {
            mv.setViewName("pages/super-groups");
        } else {
            mv.setViewName("index");
            mv.addObject("page", "pages/super-groups");
        }

        mv.addObject("superGroups", superGroups);

        return mv;
    }

    @GetMapping("/super-groups/{id}")
    public ModelAndView getSuperGroup(@RequestHeader(value = "HX-Request", required = false) boolean htmxRequest, @PathVariable("id") UUID id) {
        Optional<SuperGroupFacade.SuperGroupDTO> superGroup = this.superGroupFacade.get(id);

        if(superGroup.isEmpty()) {
            throw new RuntimeException();
        }

        ModelAndView mv = new ModelAndView();
        if(htmxRequest) {
            mv.setViewName("pages/super-group-details");
        } else {
            mv.setViewName("index");
            mv.addObject("page", "pages/super-group-details");
        }

        mv.addObject("superGroup", superGroup.get());

        return mv;
    }

    public record UpdateSuperGroupForm(int version,
                                       String name,
                                       String prettyName,
                                       String type,
                                       String svDescription,
                                       String enDescription) {
    }

    @GetMapping("/super-groups/{id}/edit")
    public ModelAndView getSuperGroupEdit(@RequestHeader(value = "HX-Request", required = true) boolean htmxRequest, @PathVariable("id") UUID id) {
        Optional<SuperGroupFacade.SuperGroupDTO> superGroup = this.superGroupFacade.get(id);

        if (superGroup.isEmpty()) {
            throw new RuntimeException();
        }

        ModelAndView mv = new ModelAndView();
        mv.setViewName("partial/edit-super-group");

        SuperGroupFacade.SuperGroupDTO sg = superGroup.get();

        mv.addObject("superGroupId", sg.id());
        mv.addObject("form", new UpdateSuperGroupForm(
                sg.version(),
                sg.name(),
                sg.prettyName(),
                sg.type(),
                sg.svDescription(),
                sg.enDescription()
        ));
        mv.addObject("types", this.superGroupFacade.getAllTypes());

        return mv;
    }

    @PutMapping("/super-groups/{id}")
    public ModelAndView updateSuperGroup(@RequestHeader(value = "HX-Request", required = false) boolean htmxRequest,
                            @PathVariable("id") UUID id,
                            final UpdateSuperGroupForm form,
                            final BindingResult bindingResult) {
        try {
            this.superGroupFacade.updateSuperGroup(
                    new SuperGroupFacade.UpdateSuperGroup(
                            id,
                            form.version,
                            form.name,
                            form.prettyName,
                            form.type,
                            form.svDescription,
                            form.enDescription
                    )
            );
        } catch (SuperGroupRepository.SuperGroupNotFoundException e) {
            throw new RuntimeException(e);
        }

        ModelAndView mv = new ModelAndView();
        mv.setViewName("pages/super-group-details");
        mv.addObject("superGroup", new SuperGroupFacade.SuperGroupDTO(
                id,
                form.version + 1,
                form.name,
                form.prettyName,
                form.type,
                form.svDescription,
                form.enDescription
        ));

        return mv;
    }

}

File: app/src/main/kotlin/it/chalmers/gamma/SuperGroupController.kt
// Spring response and form-binding signatures stay explicit at the MVC boundary.
@file:Suppress("ForbiddenVoid")

package it.chalmers.gamma

import it.chalmers.gamma.organization.CreateSuperGroup
import it.chalmers.gamma.organization.DeleteSuperGroup
import it.chalmers.gamma.organization.LocalizedText
import it.chalmers.gamma.organization.NewSuperGroup
import it.chalmers.gamma.organization.OrganizationName
import it.chalmers.gamma.organization.OrganizationQueries
import it.chalmers.gamma.organization.PrettyName
import it.chalmers.gamma.organization.SuperGroupId
import it.chalmers.gamma.organization.SuperGroupType
import it.chalmers.gamma.organization.SuperGroupUpdate
import it.chalmers.gamma.organization.UpdateSuperGroup
import it.chalmers.gamma.organization.views.renderSuperGroupDetails
import it.chalmers.gamma.organization.views.renderSuperGroupEditor
import it.chalmers.gamma.organization.views.renderSuperGroups
import jakarta.servlet.http.HttpServletRequest
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.security.web.csrf.CsrfToken
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RestController

@RestController
class SuperGroupController(
    private val organizations: OrganizationQueries,
    private val superGroupCreation: CreateSuperGroup,
    private val superGroupUpdates: UpdateSuperGroup,
    private val superGroupDeletion: DeleteSuperGroup,
) {
    @GetMapping("/super-groups", produces = [MediaType.TEXT_HTML_VALUE])
    fun superGroups(
        authentication: Authentication,
        csrfToken: CsrfToken,
        request: HttpServletRequest,
    ) = renderSuperGroups(
        pageContext(authentication, csrfToken, request),
        organizations.listSuperGroups(),
    )

    @GetMapping("/super-groups/create", produces = [MediaType.TEXT_HTML_VALUE])
    fun createSuperGroupPage(
        authentication: Authentication,
        csrfToken: CsrfToken,
        request: HttpServletRequest,
    ) = renderSuperGroupEditor(pageContext(authentication, csrfToken, request), organizations.listSuperGroupTypes())

    @PostMapping("/super-groups")
    fun createSuperGroup(
        authentication: Authentication,
        @ModelAttribute form: CreateSuperGroupForm,
    ): ResponseEntity<Void> {
        val id =
            superGroupCreation.create(
                authentication.actor(),
                NewSuperGroup(
                    OrganizationName(form.name),
                    PrettyName(form.prettyName),
                    SuperGroupType(form.type),
                    LocalizedText.of(form.svDescription, form.enDescription),
                ),
            )
        return redirect("/super-groups/${id.value}")
    }

    @GetMapping("/super-groups/{superGroupId}", produces = [MediaType.TEXT_HTML_VALUE])
    fun superGroupDetails(
        authentication: Authentication,
        csrfToken: CsrfToken,
        request: HttpServletRequest,
        @PathVariable superGroupId: String,
    ): ResponseEntity<String> {
        val details =
            organizations.superGroupDetails(SuperGroupId.parse(superGroupId))
                ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(
            renderSuperGroupDetails(
                pageContext(authentication, csrfToken, request),
                details.superGroup,
                details.groups,
            ),
        )
    }

    @GetMapping("/super-groups/{superGroupId}/edit", produces = [MediaType.TEXT_HTML_VALUE])
    fun editSuperGroup(
        authentication: Authentication,
        csrfToken: CsrfToken,
        request: HttpServletRequest,
        @PathVariable superGroupId: String,
    ): ResponseEntity<String> {
        val editor =
            organizations.superGroupEditor(SuperGroupId.parse(superGroupId))
                ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(
            renderSuperGroupEditor(
                pageContext(authentication, csrfToken, request),
                editor.superGroupTypes,
                editor.superGroup,
            ),
        )
    }

    @GetMapping("/super-groups/{superGroupId}/cancel-edit", produces = [MediaType.TEXT_HTML_VALUE])
    fun cancelSuperGroupEdit(
        authentication: Authentication,
        csrfToken: CsrfToken,
        request: HttpServletRequest,
        @PathVariable superGroupId: String,
    ): ResponseEntity<String> = superGroupDetails(authentication, csrfToken, request, superGroupId)

    @PutMapping("/super-groups/{superGroupId}")
    fun updateSuperGroup(
        authentication: Authentication,
        @PathVariable superGroupId: String,
        @ModelAttribute form: UpdateSuperGroupForm,
    ): ResponseEntity<Void> {
        superGroupUpdates.update(
            authentication.actor(),
            SuperGroupUpdate(
                SuperGroupId.parse(superGroupId),
                form.version,
                OrganizationName(form.name),
                PrettyName(form.prettyName),
                SuperGroupType(form.type),
                LocalizedText.of(form.svDescription, form.enDescription),
            ),
        )
        return redirect("/super-groups/$superGroupId")
    }

    @DeleteMapping("/super-groups/{superGroupId}")
    fun deleteSuperGroup(
        authentication: Authentication,
        @PathVariable superGroupId: String,
    ): ResponseEntity<Void> {
        superGroupDeletion.delete(authentication.actor(), SuperGroupId.parse(superGroupId))
        return redirect("/super-groups")
    }
}

data class CreateSuperGroupForm(
    val name: String,
    val prettyName: String,
    val type: String,
    val svDescription: String,
    val enDescription: String,
)

data class UpdateSuperGroupForm(
    val name: String,
    val prettyName: String,
    val type: String,
    val svDescription: String,
    val enDescription: String,
    val version: Int,
)
